syntax = "proto2";
package protos;

import "protos/hyperparams.proto";

message Convnet {
  oneof convnet_oneof {
    CrnnNet crnn_net = 1;
    ResNet resnet = 2;
    StnConvnet stn_convnet = 3;
    TgnDeconvnet tgn_deconvnet = 4;
  }
}

message CrnnNet {
  enum NetType {
    SINGLE_BRANCH = 0;
    TWO_BRANCHES = 1;
    THREE_BRANCHES = 2;
  }
  optional NetType net_type = 1 [default = SINGLE_BRANCH];
  optional Hyperparams conv_hyperparams = 2;
  optional bool summarize_activations = 3 [default=false];
  optional bool tiny = 4 [default = false];
}

message ResNet {
  enum NetType {
    SINGLE_BRANCH = 0;
    TWO_BRANCHES = 1;
    THREE_BRANCHES = 2;
  }
  enum NetDepth {
    RESNET_30 = 0;
    RESNET_50 = 1;
    RESNET_100 = 2;
  }
  optional NetType net_type = 1 [default = SINGLE_BRANCH];
  optional NetDepth net_depth = 2 [default = RESNET_50];
  optional Hyperparams conv_hyperparams = 3;
  optional bool summarize_activations = 4 [default=false];
}

message StnConvnet {
  optional Hyperparams conv_hyperparams = 1;
  optional bool summarize_activations = 2 [default=false];
  optional bool tiny = 3 [default = false];
}

message TgnDeconvnet {
  optional Hyperparams conv_hyperparams = 1;
  optional bool summarize_activations = 2 [default=false];
}

